[yuqing-tham] iP

yuqing-tham opened the pull request:
# Universe 

> "Could it be magic?" - Yuqi, (G)I-DLE

Universe helps you keep track of the Chores you need to complete. 
Universe is 
- text-based
- beginner-friendly
- ~~efficient~~ **SUPER** efficient 
- most importantly, it's <ins>FREE</ins> 🔥 

## Download instructions
Follow the steps below to run Universe: 
1. Download it [here](https://github.com/yuqing-tham/ip/releases/tag/A-Release)
2. Once the JAR file is downloaded, please `cd` to the correct folder where the file is located. 
3. Run the following command `java -jar Universe.jar`. 
4. You're all set to start a personal Cosmic Chore Checklist! 

## Features
- [x] Add ToDos (prepping for your tutorial, perhaps?)
- [x] Add Deadlines (_maybe_ an assignment with a due date)
- [x] Add Events (a meeting?)
- [x] Mark the Chores as done or not done
- [x] Remove the Chores from the Checklist 
- [x] Filter the Chores in the Checklist by date
- [x] Find Chores with the keyword
- [ ] Stay tuned for new features 😉  

## A peek into the workings of the Universe 
The `main` method: 
```ruby 
public static void main(String[] args) {
        new Universe("data/chores.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

yuqing-tham committed:
Add package UNIverse, Enter, Exit and UNIverse classes

yuqing-tham committed:
Add interaction functionality and new greetings

yuqing-tham committed:
Add comments

yuqing-tham committed:
Add Checklist function

yuqing-tham committed:
Add Chore class and Mark Chore as Done function

yuqing-tham committed:
Add todo, deadline and event functionality

yuqing-tham committed:
Edit commands

yuqing-tham committed:
Add Automated Text UI Testing

yuqing-tham committed:
Add exceptions

yuqing-tham committed:
Add Delete Chore function

yuqing-tham committed:
Add previously uncaught Exceptions

yuqing-tham committed:
Rename package and class to comply with coding standard

yuqing-tham committed:
Reformat switch indentation

yuqing-tham committed:
Fix formatting issues

yuqing-tham committed:
Chore class: Add newline after print chore

yuqing-tham committed:
Update Ui Testing

yuqing-tham committed:
Add read and save to file functionality

yuqing-tham committed:
Merge Level 7 from branch-Level-7

yuqing-tham committed:
Handle file does not exist

yuqing-tham committed:
Read deadline date as LocalDate

yuqing-tham committed:
Handle DateTimeParseException

yuqing-tham committed:
Fix deadline and event date/time formatting

yuqing-tham committed:
Add filter date function

yuqing-tham committed:
Merge Level 8 from branch-Level-8

yuqing-tham committed:
Rearrange code to make it more OOP

yuqing-tham committed:
Fix ResponseManager and Universe class

yuqing-tham committed:
Rearrange files into packages

yuqing-tham committed:
Merge branch 'add-gradle-support'

yuqing-tham committed:
Setup build.gradle for build and run using Gradle

yuqing-tham committed:
Add JUnit Tests for some classes

yuqing-tham committed:
Add JUnit tests for classes under the package response

yuqing-tham committed:
Reconfigure build.gradle file for JAR generation

yuqing-tham committed:
Add JavaDoc comments

yuqing-tham committed:
Cleanup code according to coding standard

yuqing-tham committed:
Add find functionality

yuqing-tham committed:
Merge branch-A-JavaDoc

yuqing-tham committed:
Merge branch-A-CodingStandard

yuqing-tham committed:
Resolve merge conflicts when merging Level 9

yuqing-tham committed:
Add Parser JUnit test cases

yuqing-tham committed:
Add checkstyle configurations

Zhannyhong commented on src/main/java/universe/Checklist.java:
```diff
@@ -0,0 +1,141 @@
+     * @param date the date to filter in String form
+     * @param formattedDate the date to filter in LocalDateTime form
+     */
+    public void filter(String date, LocalDateTime formattedDate) {
```
Ambiguous method name. Perhaps rename to something where the user knows what it is filtering by?

Zhannyhong commented on src/main/java/universe/Checklist.java:
```diff
@@ -0,0 +1,141 @@
+     * Finds the chores in the checklist that contains the keyword and prints them
+     * @param keyword to be checked against the chore descriptions
+     */
+    public void find(String keyword) {
```
Similarly ambiguous method name. Perhaps rename to something more meaningful

Zhannyhong commented on src/main/java/universe/exceptions/EmptyResponseException.java:
I like how you extended the UniverseException to create various exception subtypes

Zhannyhong commented on src/main/java/universe/response/CheckResponse.java:
```diff
@@ -0,0 +1,82 @@
+     * @throws InvalidResponseException if user inputs a unrecognisable command
+     * @throws MissingListItemException if user inputs a command with a list number not in Checklist
+     */
+    public void handleError() throws EmptyResponseException, MissingDescriptionException, MissingTimeException,
```
If throwing so many exceptions, perhaps can just use `throws UniverseException` instead?

Zhannyhong commented on src/main/java/universe/response/ResponseManager.java:
```diff
@@ -0,0 +1,75 @@
+        Parser r = new Parser(response);
+
+        if (response.contains("list")) { // prints checklist to screen if user says "list"
+            System.out.println("Cosmic Chore Checklist:");
+            list.printChecklist();
+        } else if (response.startsWith("check")) { // mark chore as done
+            int choreNumber = r.getChoreNumber();
+            list.checkAsDone(choreNumber);
+        } else if (response.startsWith("uncheck")) { // mark chore as not done
+            int choreNumber = r.getChoreNumber();
+            list.uncheckAsDone(choreNumber);
+        } else if (response.startsWith("todo")) { // deals with a todo chore
+            String description = r.getDescription();
+            ToDo todo = new ToDo(description);
+            list.addChore(todo);
+        } else if (response.startsWith("deadline")) { // deals with a deadline chore
+            String description = r.getDeadlineDescription();
+            LocalDateTime date = r.getDate();
+            Deadline deadline = new Deadline(description, date);
+            list.addChore(deadline);
+        } else if (response.startsWith("event")) { // deals with an event chore
+            String description = r.getEventDescription();
+            LocalDateTime start = r.getStartTime();
+            LocalDateTime end = r.getEndTime();
+            Event event = new Event(description, start, end);
+            list.addChore(event);
+        } else if (response.startsWith("remove")) { // deals with removing
+            int choreNumber = r.getChoreNumber();
+            list.removeChore(choreNumber);
+        } else if (response.startsWith("filter")) { // deals with filtering all the chores on a date
+            String dateString = r.getFilterDateString();
+            LocalDateTime date = r.getFilterDate();
+            list.filter(dateString, date);
+        } else if (response.startsWith("find")) { // deals with finding chores with the keyword
+            String keyword = r.getKeyword();
+            list.find(keyword);
+        } else { // command not recognised, so an invalid response exception is thrown
+            throw new InvalidResponseException();
+        }
+    }
```
Since there are so many if/else blocks, perhaps try to simplify with a switch case instead?

Zhannyhong reviewed (commented):
LGTM. Not much naming issues!

wilsonh0 commented on src/main/java/universe/Checklist.java:
```diff
@@ -0,0 +1,141 @@
+ * @author yuqing-tham
+ */
+public class Checklist {
+    private static ArrayList<Chore> checklist = new ArrayList<>(100);
```
The ArrayList checklist represents a collection of Chore objects, so it should use a plural form for better readability. Could possibly rename it to chores to make it clearer that it holds multiple items.

yuqing-tham committed:
Add GUI

yuqing-tham committed:
Merge branch 'branch-Level-10'

yuqing-tham committed:
Fix Suppress Warning issue

wilsonh0 commented on src/main/java/universe/response/ResponseManager.java:
```diff
@@ -0,0 +1,75 @@
+     * Asks a Parser instance to help split the response to meaningful parts.
+     * @throws InvalidResponseException if user input not recognised under list of keywords
+     */
+    public void execute() throws InvalidResponseException {
```
I think a switch statement could help reduce visual clutter and improve readability in the execute() method. The current long if-else chain makes it slightly harder to follow, and a switch would make it clearer.

wilsonh0 reviewed (commented):
I think you did a great job following the coding standards! Everything looks well-structured, and any issues I found were just minor nitpicks. One small suggestion would be to add more separation between logical blocks to improve readability.

yuqing-tham committed:
Fix list, filter and find commands bug

The list, filter and find commands are not recognized by the GUI.

Let's modify the error handling function to account for list, filter and
find commands. Doing so will resolve the bug.

yuqing-tham committed:
Fix unable to read file bug for JAR file

To resolve the issue of missing chores.txt file when running the JAR
file GUI.

yuqing-tham committed:
Rename some variables and functions

Some variable and function names are not descriptive enough.

Let's change the naming of the variables and functions to better
describe what it is about.

yuqing-tham committed:
Add assertions to document important assumptions

There are certain parts of the code that work under certain assumptions.

Let's document these assumptions by adding assertions.

yuqing-tham committed:
Modify code for better readability

yuqing-tham committed:
Merge pull request #1 from yuqing-tham/branch-A-Assertions

Branch A-Assertions

yuqing-tham committed:
Resolve merge conflicts

yuqing-tham committed:
Merge pull request #2 from yuqing-tham/branch-A-CodeQuality

Branch A-CodeQuality

yuqing-tham committed:
Add Clear All function for C-MassOps

yuqing-tham committed:
Merge pull request #3 from yuqing-tham/branch-BCD-Extension

Branch-C-MassOps

yuqing-tham committed:
Change Universe Picture

yuqing-tham committed:
Change header comments to follow coding standard

yuqing-tham committed:
Add support to manage hi and help commands

yuqing-tham committed:
Change App Icon and Title

yuqing-tham committed:
Add support for window resizing

yuqing-tham committed:
Change background and dialog box colours

yuqing-tham committed:
Change user profile picture

yuqing-tham committed:
Merge pull request #5 from yuqing-tham/branch-A-BetterGui

Merge Branch A-BetterGui

yuqing-tham committed:
Update header comments for ResponseChecker class

yuqing-tham committed:
Handle endDateTime before startDateTime error

yuqing-tham committed:
Fix variable naming to follow coding standard

yuqing-tham committed:
Add User Guide in README file

yuqing-tham committed:
Resolve UI screenshot issue

yuqing-tham committed:
Rewrite code to follow SLAP

yuqing-tham committed:
Add Javadoc comments to new methods

yuqing-tham committed:
Clean up code

yuqing-tham committed:
Update link to JAR file in README

yuqing-tham committed:
Fix issues in code

yuqing-tham committed:
Update README to include examples

yuqing-tham committed:
Add command to exit GUI support

yuqing-tham committed:
Update README to include bye command

yuqing-tham committed:
Add Javadoc for test classes

yuqing-tham committed:
Fix Javadoc for test classes

yuqing-tham committed:
Fix formatting error in Javadoc

yuqing-tham committed:
Fix catch Exception blocks

yuqing-tham committed:
Remove general exception catch block

yuqing-tham committed:
Remove unused import

yuqing-tham committed:
Update link to JAR file in README

yuqing-tham committed:
Update README.md
